src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs

```diff
@@ -0,0 +1,185 @@
+                if (string.Equals(enabledPreviewFeatures, "true", System.StringComparison.OrdinalIgnoreCase))
+                {
+                    return;
+                }
+
+                ConcurrentDictionary<ISymbol, IOperation?> requiresPreviewFeaturesSymbols = new ConcurrentDictionary<ISymbol, IOperation?>();
+                context.RegisterOperationAction(context => BuildSymbolInformation(context, requiresPreviewFeaturesSymbols),
+                    OperationKind.Invocation,
+                    OperationKind.ObjectCreation,
+                    OperationKind.PropertyReference,
+                    OperationKind.FieldReference,
+                    OperationKind.DelegateCreation,
+                    OperationKind.EventReference
+                    );
+            });
+        }
+
+        private void BuildSymbolInformation(OperationAnalysisContext context, ConcurrentDictionary<ISymbol, IOperation?> requiresPreviewFeaturesSymbols)
+        {
+            if (OperationUsesPreviewFeatures(context.Operation, requiresPreviewFeaturesSymbols))
+            {
+                context.ReportDiagnostic(context.Operation.CreateDiagnostic(Rule));
+            }
+        }
+
+        private bool OperationUsesPreviewFeatures(IOperation operation, ConcurrentDictionary<ISymbol, IOperation?> requiresPreviewFeaturesSymbols)
+        {
+            ISymbol? symbol = GetOperationSymbol(operation);
+            if (symbol == null)
+            {
+                return false;
+            }
+
+            return OperationUsesPreviewFeatures(operation, symbol, requiresPreviewFeaturesSymbols, true);
+        }
+
+        private static ISymbol? GetOperationSymbol(IOperation operation)
+            => operation switch
+            {
+                IInvocationOperation iOperation => iOperation.TargetMethod,
+                IObjectCreationOperation cOperation => cOperation.Constructor,
+                IPropertyReferenceOperation pOperation => pOperation.Property,
+                IFieldReferenceOperation fOperation => fOperation.Field,
+                IDelegateCreationOperation dOperation => dOperation.Type,
+                IEventReferenceOperation eOperation => eOperation.Member,
+                _ => null,
+            };
+
+        private bool TryGetOrCachePreviewFeaturesAttributesOnSymbol(IOperation operation, ISymbol symbol, ConcurrentDictionary<ISymbol, IOperation?> requiresPreviewFeaturesSymbols, bool checkParents)
+        {
+            if (!requiresPreviewFeaturesSymbols.TryGetValue(symbol, out IOperation? existing))
+            {
+                if (checkParents)
+                {
+                    var containingSymbol = symbol.ContainingSymbol;
+                    // Namespaces do not have attributes
+                    while (containingSymbol is INamespaceSymbol)
+                    {
+                        containingSymbol = containingSymbol.ContainingSymbol;
+                    }
```

Comment:
Can a namespace be contained in a symbol that's not a namespace?
i.e, if we have a namespace here, won't we be getting ContainingSymbol and eventually will end up with null?

Reply:
>eventually will end up with null?

Indeed. That is the behavior I want here. The [RequiresPreviewFeature] attribute can be set as an assembly level attribute, so I need to go up the symbol chain all the way to the root. So a method like `System.Text.SomeMethod()`, will have the chain `SomeMethod -> Text -> System (this is the namespace) -> Current Module -> Current Assembly -> null.